### DISCUSSION TITLE
JapReader occasionally doesn't read text from the clipboard (when textractor is running)

### DISCUSSION THREAD

**Discussion by benissimus** [marisukukise/japReader#34] [General]
I've been using japreader together with textractor on my laptop for a while and it works fine, but recently I got a PC and on that PC japReader will sometimes ignore Japanese text in the clipboard. After some testing and comparing the settings on the laptop and the PC I was able to confirm that this only happens when Textractor is also open, so I can't say for sure if this is a japReader issue. It affects both the text copied from textractor and outside of it, for example from a web browser. Just to be clear, the text is copied successfully and I can paste it, but japReader doesn't process it. 
Laptop has Win 10 home, PC Win 10 Pro and pretty much the same language settings.
Are there any settings in japReader I could tinker with to get around this?

**Comment by marisukukise** [comment]
I tried looking into it, and sometimes the text copied from the clipboard is converted into question marks (and subsequently it's not getting parsed by ichi.moe).
But it's really hard to reproduce, I only got it like 3-4 times in about 2 hours, so I'm still not sure what the reason is. It only happens very soon after clipboard change, because if I read from the clipboard twice (one line of code below another) then the first one might be garbled, but the second one not. I need to look into it more.

**Reply by benissimus** [reply]
It happens fairly consistently on my end so let me know if I could gather any useful info

**Reply by marisukukise** [reply]
> It happens fairly consistently on my end so let me know if I could gather any useful info

I created a new branch https://github.com/marisukukise/japReader/tree/clipboard-issue

If you wanna gather some info:
1. Clone the new branch `git clone -b clipboard-issue https://github.com/marisukukise/japReader.git`
2. Enter the dir `cd japReader`
3. Install dependencies `npm i` (install node.js if you don't have it)
4. Run japReader `npm start`
5. There should be a new window "japReader - Clipboard Handler". Focus it and press Ctrl+Shift+I, and then "Console"
6. Keep using japReader as usual, and if something works wrong, then make a screenshot of the Console output and also the original copied sentence if it's not shown in the console already. And the source of the copied sentence.

I want to know if only specific sentences are getting copied wrong, or any random sentences. And also if it's related to Textractor or not.

**Comment by benissimus** [comment]
1. This only happens when Textractor is open. It doesn't matter if copy to clipboard extension is enabled in Textractor
2. It happens to random sentences, if I copy a sentence several times it'll eventually get picked up
Three screenshots taken when copying [Momotaro](https://en.wikibooks.org/wiki/Japanese/Reader/Momotaro) from wikibooks, the last one is from the game. Lines that are empty under "clipboard contents (1)" do not get displayed in japReader
[screenshots.zip](https://github.com/marisukukise/japReader/files/11724541/screenshots.zip)

**Reply by marisukukise** [reply]
Thanks for testing
I'm testing it further and I also occasionally get it, but very rarely (2 times in about 5000 copy operations)
Also it doesn't always happen on the first read as I thought, but on random reads as well (for example on the sixth consecutive read without any other operations in between). 
![image](https://github.com/marisukukise/japReader/assets/109931938/df71cf49-9c14-4083-b476-49a4d1cfc119)

I'm still trying to confirm if it can happen without Textractor running, but for now I can ask you, are you using HDD or SSD, and also how much RAM do you have? Maybe it's related to disk speed. 

Either way for now a quick fix would be to read the clipboard a few times in a row and choose the value that is in majority, because the corrupted ones should be minority. It's kind of an ugly solution and I still don't know the main reason it's happening, but if I don't figure it out soon I'll probably implement this quick fix.

**Reply by benissimus** [reply]
16 GB RAM, NVMe SSD. The laptop where it works fine also has an SSD but it's much slower. This probably doesn't make any sense, but could japReader be reading the clipboard too early?

**Reply by marisukukise** [reply]
I don't think it reads too early, because the [clipboard-event](https://www.npmjs.com/package/clipboard-event) just detects the clipboard change, it can't detect it before the clipboard was changed. And the reading of clipboard contents is done through Electron JS which the entire app is written in. It mostly works as intended but apparently less than 1% of reads sometimes have "bursts" of errors when reading repeatedly. It could just be some memory leaks or for all I know, it could be cosmic rays passing through your CPU... (joke). I tested it on 1_300_000 readings for each configuration and Textractor definitely increases the number of misreads but it happens without it too. 
![image](https://github.com/marisukukise/japReader/assets/109931938/8722ef97-4b15-40f9-89d4-47bfc372dd5a)

I added a previously mentioned fix to the https://github.com/marisukukise/japReader/tree/develop branch. It now reads the clipboard 11 (number chosen for performance after some tests) times and finds the most common string. So it doesn't completely solve the issue, because it can still happen that 6 times out of 11 will read the string incorrectly, but it adds robustness and it's much less likely to misread. 

So if you have a moment you can test japReader on that branch and see if it's better. The console logs are still available, so if something's still not right you can check the outputs. And if it works well enough I will add it in 1.9.1

**Reply by benissimus** [reply]
The development branch works fine, there are some errors in the console but it didn't skip any lines
![dev](https://github.com/marisukukise/japReader/assets/136090341/83479849-5c99-4255-b356-fd9e88a3d0e3)

**Comment by marisukukise** [comment]
Fixed to the best of my ability in https://github.com/marisukukise/japReader/releases/tag/v1.9.1

But I also changed the method a bit, so if the problem persists open a new issue
